### PR TITLE
Benchmark payload

### PR DESCRIPTION
**Pull request by francois141**

### CODE REVIEW

**Comment by francois141** [comment]
Ready to merge mon capitaine!

File: config/test/spike-virt-benchmark.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# A simple configuration to run on QEMU virt platform
```
Comment (CharlyCst) [inline config/test/spike-virt-benchmark.toml:1]: Small nits, but we should try to update the top-level comment to explain what the config does :)

File: firmware/tracing_firmware/main.rs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        let mcounteren: u32;
+        asm!("csrr {}, mcounteren", out(reg) mcounteren);
+        asm!("csrw mcounteren, {}", in(reg) mcounteren | 1);
+    }
+}
+
+fn main() -> ! {
+    unsafe {
+        asm!("la sp, 0x80700000");
+    }
```
Comment (CharlyCst) [inline firmware/tracing_firmware/main.rs:21]: You are trying to setup the stack of the firmware here right?
The `setup_binary!` macro already creates a proper stack, so this should not be needed, at least not for the firmware.
The OS will need its own stack though.

File: firmware/tracing_firmware/main.rs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+
+        os = in(reg) os,
+        mtvec = in(reg) trap,
+        mpp = in(reg) mpp,
+        out("t4") _,
+        );
+    }
+
+    success();
```
Comment (CharlyCst) [inline firmware/tracing_firmware/main.rs:49]: If the OS is supposed to exit I would suggest to use `failure()` here, as it means we failed to jump into the OS.

File: firmware/tracing_firmware/main.rs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    r#"
+.text
+.align 4
+.global _empty_handler
+_empty_handler:
+    // Skip illegal instruction (pc += 4)
+    csrrw x5, mepc, x5
+    addi x5, x5, 4
+    csrrw x5, mepc, x5
+    // Return back to miralis
```
Comment (CharlyCst) [inline firmware/tracing_firmware/main.rs:64]: It is not going back to Miralis, but to the OS here.

File: firmware/tracing_firmware/main.rs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    success();
+}
+
+fn trigger_ctx_switch_to_firmware() -> usize {
+    let begin: u64;
+    let end: u64;
+
+    unsafe {
+        // Read the `mcycle` register (assuming 64-bit RISC-V)
+        asm!("csrr {}, cycle", out(reg) begin);
```
Comment (CharlyCst) [inline firmware/tracing_firmware/main.rs:124]: The only thing we need to be sure here is that reading the cycles doesn't trap :)

You were also talking about taking one big measurement of `NB_REPEATS` traps in a row to double check, do you want to do it here too or do we keep that for future work?

File: firmware/tracing_firmware/main.rs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+extern "C" {
+    fn _empty_handler();
+}
+
+// —————————————————————————————— Benchmark operating system —————————————————————————————— //
+
+const NB_REPEATS: usize = 1000;
+
+pub fn bubble_sort(arr: &mut [usize; NB_REPEATS]) {
```
Comment (CharlyCst) [inline firmware/tracing_firmware/main.rs:77]: Nice haha

**Pull request merged**